base/accumulate.jl

```diff
@@ -0,0 +1,450 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+"""
+    Accumulate(op[, v0], iter)
+
+An iterator which cumulatively applies the binary operation `op` to the iterator `iter`, in a similar manner
+to [`foldl`](@ref) but returning intermediate values.
+
+If an initial `v0` value is provided, then the first value of the iterator will be
+
+    op(v0, first(iter))
+
+Otherwise it will be
+
+    Base.reduce_first(op, first(iter))
+
+See also [`Base.reduce_first`](@ref) and [`accumulate`](@ref).
+"""
+struct Accumulate{O,V,I}
+    op::O
+    v0::V
+    iter::I
+end
+Accumulate(op, iter) = Accumulate(op, uninitialized, iter) # use `uninitialized` as a sentinel
+
+# the following is largely based on Generator objects
+Base.IteratorEltype(acc::Accumulate) = EltypeUnknown()
+Base.IteratorSize(acc::Accumulate) = Base.IteratorSize(acc.iter)
+
+length(acc::Accumulate) = length(acc.iter)
+size(acc::Accumulate)   = size(acc.iter)
+axes(acc::Accumulate)   = axes(acc.iter)
+ndims(acc::Accumulate)  = ndims(acc.iter)
+
+start(acc::Accumulate) = (@_inline_meta; (start(acc.iter), acc.v0))
+done(acc::Accumulate, accstate) = (@_inline_meta; done(acc.iter, accstate[1]))
+function next(acc::Accumulate, accstate)
+    @_inline_meta
+    itrstate, accval = accstate
+    val, itrstate = next(acc.iter, itrstate)
+    if accval === uninitialized
+        accval = reduce_first(acc.op, val)
```

Comment:
So, this is type-unstable even for things like `Accumulate(+, itr)`?

Reply:
Yes, but the new Union optimisations mean that there is no performance hit.

Reply:
As I understand it, the `Union` optimizations _help_ with cases like this, but retaining type stability will still be faster. Worth Nanosoldiering or otherwise benchmarking to be sure though.

Reply:
In most cases, the only type instability will be in the first call to `next`, which is typically unrolled, so we should be able to avoid that.